### PR TITLE
Add /auth/tokens endpoint and auth store to framework

### PR DESCRIPTION
**Pull request by eamansour**
## Why?
See https://github.com/galasa-dev/projectmanagement/issues/1805

## Changes
- Added a new `/auth/tokens` endpoint to list tokens from the back-end auth store, sorted by earliest-first.
  - Updated the OpenAPI docs to include this new endpoint
- Added a new auth store to the framework, which is only initialised when the `framework.auth.store` CPS property is set.
- Added logic to store a new token into the auth store when a user is creating a new access token (i.e. when they are not refreshing an existing token, and when they have provided a description of the token to be created)
- Refactored existing framework tests to use a common MockFramework instead of creating a new mock framework class within each test suite

### CODE REVIEW

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-framework-pr-546-9m5k9 for details. If you are unable to do so, please contact a member of the Galasa team.

**Comment by galasa-team** [comment]
Build successful

File: galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/beans/TokenPayload.java [hunk 1]
```diff
@@ -23,6 +23,9 @@ public class TokenPayload {
     @SerializedName("code")
     private String code;
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/beans/TokenPayload.java:25]: move this bean to the beans bundle so we can generate it easier

File: galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/JwtWrapper.java [hunk 2]
```diff
@@ -45,6 +45,11 @@ public JwtWrapper(HttpServletRequest req, Environment env) {
         usernameClaimOverrides = env.getenv(EnvironmentVariables.GALASA_USERNAME_CLAIMS);
     }
 
+    public JwtWrapper(String jwt, Environment env) {
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/JwtWrapper.java:48]:
```    private static final List<String> DEFAULT_USERNAME_CLAIMS = List.of(
        "preferred_username",
        "name",
        "sub"
    );
    
```
We want to get rid of this. And if the user's claim names are not present then blow up with a decent error.

File: galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/routes/AuthTokensRoute.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        return getResponseBuilder().buildResponse(response, "application/json", getTokensAsJsonString(tokensToReturn), HttpServletResponse.SC_OK);
+    }
+
+    /**
+     * Converts a list of authentication tokens into a JSON string
+     *
+     * @param tokens the tokens to convert
+     * @return a JSON representation of the tokens within a "tokens" JSON array
+     */
+    private String getTokensAsJsonString(List<AuthToken> tokens) {
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/routes/AuthTokensRoute.java:80]: @NotNull annnotation

File: galasa-parent/dev.galasa.framework.api.authentication/src/test/java/dev/galasa/framework/api/authentication/routes/AuthTokensRouteTest.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+        // When...
+        servlet.init();
+        servlet.doGet(mockRequest, servletResponse);
+
+        // Then...
+        assertThat(servletResponse.getStatus()).isEqualTo(500);
+        assertThat(servletResponse.getContentType()).isEqualTo("application/json");
+        assertThat(servletResponse.getHeader("Access-Control-Allow-Origin")).isEqualTo("*");
+        assertThat(outStream.toString()).contains("GAL5053E", "Error retrieving tokens from the auth store");
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework.api.authentication/src/test/java/dev/galasa/framework/api/authentication/routes/AuthTokensRouteTest.java:147]: Better user-centered error message pls ?

File: galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java [hunk 5]
```diff
@@ -84,7 +84,11 @@ public enum ServletErrorMessage {
 
     // Auth APIs...
     GAL5051_INVALID_GALASA_TOKEN_PROVIDED             (5051, "E: Invalid GALASA_TOKEN value provided. Please ensure you have set the correct GALASA_TOKEN property for the targeted ecosystem at ''{0}'' and try again."),
-    GAL5052_FAILED_TO_RETRIEVE_CLIENT                 (5052, "E: Unable to retrieve client for authentication. Please ensure you have set the correct GALASA_TOKEN property for the targeted ecosystem at ''{0}'' and try again.")
+    GAL5052_FAILED_TO_RETRIEVE_CLIENT                 (5052, "E: Unable to retrieve client for authentication. Please ensure you have set the correct GALASA_TOKEN property for the targeted ecosystem at ''{0}'' and try again."),
+    GAL5053_FAILED_TO_RETRIEVE_TOKENS                 (5053, "E: Error retrieving tokens from the auth store. Report the problem to your Galasa Ecosystem owner."),
+    GAL5054_FAILED_TO_GET_CONNECTOR_URL               (5054, "E: Could not get the URL of an upstream identity provider to authenticate with. Report the problem to your Galasa Ecosystem owner."),
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java:89]: Internal server error. The REST API server could not get the URL of the authentication provider (eg: github/oidc) from the dex Galasa component. The dex component of Galasa could be badly configured, or could be experiencing a temporary issue. Report the problem to your Galasa Ecosystem owner.

File: galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/SetupEcosystem.java [hunk 6]
```diff
@@ -11,87 +11,73 @@
...
 public class SetupEcosystem {
-    
+
     private Log logger = LogFactory.getLog(this.getClass());
 
     private static final String DSS_ENV_VAR   = "GALASA_DYNAMICSTATUS_STORE";
     private static final String RAS_ENV_VAR   = "GALASA_RESULTARCHIVE_STORE";
     private static final String CREDS_ENV_VAR = "GALASA_CREDENTIALS_STORE";
-    
+    private static final String AUTH_ENV_VAR  = "GALASA_AUTH_STORE";
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/SetupEcosystem.java:27]: What is this used for ?
Given we don't support Docker deployments now, do we need this ?

**Review by techcobweb — APPROVED** [review]
Approved pending comments being resolved.

File: galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/beans/TokenPayload.java (hunk 1 above)
Comment (eamansour) [reply galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/beans/TokenPayload.java:25]: Done

File: galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java (hunk 5 above)
Comment (eamansour) [reply galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java]: Done in https://github.com/galasa-dev/framework/pull/546/commits/e69cef6b4144a0cf1e64dd35b27af1e642be9ca9

**Comment by galasa-team** [comment]
Build successful

File: galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/SetupEcosystem.java (hunk 6 above)
Comment (eamansour) [reply galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/SetupEcosystem.java:27]: This is invoked when running the boot jar with the `--setupeco` flag, and is used in the Helm chart to set the `framework.xxx.store` properties into the CPS. Would be great to remove if it's not needed, but that would likely need to be a separate story.

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**